File: app/src/main/java/pl/laskowski/marcin/solving/deduction/solver/DeductionSolverV3.java
package pl.laskowski.marcin.solving.deduction.solver;

import pl.laskowski.marcin.solving.deduction.algorithm.*;
import pl.laskowski.marcin.model.type.SudokuType;

import java.util.ArrayList;
import java.util.List;

public class DeductionSolverV3 extends DeductionSolver {

    @Override
    protected List<DeductionAlgorithm.Factory> provideAlgorithms(SudokuType type) {
        List<DeductionAlgorithm.Factory> algorithms = new ArrayList<>();
        algorithms.add(new HintEliminationAlgorithm.Factory());
        algorithms.add(new NakedSinglesAlgorithm.Factory());
        algorithms.add(new HiddenSinglesAlgorithm.Factory());
        for (int i = 2; i <= type.getRegionSize() / 2; i++) {
            algorithms.add(new NakedValuesAlgorithm.Factory(i));
            algorithms.add(new HiddenValuesAlgorithm.Factory(i));
        }
        algorithms.add(new RegionIntersectionAlgorithm.Factory());
        return algorithms;
    }
}

File: core/src/main/kotlin/io/github/ilikeyourhat/kudoku/solving/deduction/solver/DeductionSolverV3.kt
package io.github.ilikeyourhat.kudoku.solving.deduction.solver

import io.github.ilikeyourhat.kudoku.model.SudokuType
import io.github.ilikeyourhat.kudoku.solving.deduction.algorithm.DeductionAlgorithm
import io.github.ilikeyourhat.kudoku.solving.deduction.algorithm.HiddenSinglesAlgorithm
import io.github.ilikeyourhat.kudoku.solving.deduction.algorithm.HiddenValuesAlgorithm
import io.github.ilikeyourhat.kudoku.solving.deduction.algorithm.HintEliminationAlgorithm
import io.github.ilikeyourhat.kudoku.solving.deduction.algorithm.NakedSinglesAlgorithm
import io.github.ilikeyourhat.kudoku.solving.deduction.algorithm.NakedValuesAlgorithm
import io.github.ilikeyourhat.kudoku.solving.deduction.algorithm.RegionIntersectionAlgorithm

class DeductionSolverV3 : DeductionSolver() {

    override fun provideAlgorithms(type: SudokuType): List<DeductionAlgorithm.Factory> {
        val algorithms = mutableListOf(
            HintEliminationAlgorithm.Factory(),
            NakedSinglesAlgorithm.Factory(),
            HiddenSinglesAlgorithm.Factory()
        )
        for (i in 2..type.maxValue / 2) {
            algorithms.add(NakedValuesAlgorithm.Factory(i))
            algorithms.add(HiddenValuesAlgorithm.Factory(i))
        }
        algorithms.add(RegionIntersectionAlgorithm.Factory())
        return algorithms
    }
}
